statsmodels/tsa/statespace/sarimax.py

```diff
@@ -0,0 +1,1941 @@
+
+        # State intercept (Harvey) or additional observation intercept
+        # (Hamilton)
+        # SARIMA trend enters through the a time-varying state intercept,
+        # associated with the first row of the stationary component of the
+        # state vector (i.e. the first element of the state vector following
+        # any differencing elements)
+        if self.k_trend > 0:
+            data = np.dot(self._trend_data, params_trend).astype(params.dtype)
+            if not self.hamilton_representation:
+                self['state_intercept', self._k_states_diff, :] = data
+            else:
+                # The way the trend enters in the Hamilton representation means
+                # that the parameter is not an ``intercept'' but instead the
+                # mean of the process. The trend values in `data` are meant for
+                # an intercept, and so must be transformed to represent the
+                # mean instead
+                if self.hamilton_representation:
+                    data /= np.sum(-reduced_polynomial_ar)
+
+                # If we already set the observation intercept for MLE
+                # regression, just add to it
+                if self.mle_regression:
+                    self.obs_intercept += data[None, :]
+                # Otherwise set it directly
+                else:
+                    self.obs_intercept = data[None, :]
+
+        # Observation covariance matrix
+        if self.measurement_error:
+            self['obs_cov', 0, 0] = params_measurement_variance
+
+        # Transition matrix
+        if self.k_ar > 0 or self.k_seasonal_ar > 0:
+            self[self.transition_ar_params_idx] = reduced_polynomial_ar[1:]
+        elif not self.transition.dtype == params.dtype:
+            self.transition = self.transition.real.astype(params.dtype)
+
+        # Selection matrix (Harvey) or Design matrix (Hamilton)
+        if self.k_ma > 0 or self.k_seasonal_ma > 0:
+            if not self.hamilton_representation:
+                self[self.selection_ma_params_idx] = reduced_polynomial_ma[1:]
+            else:
+                self[self.design_ma_params_idx] = reduced_polynomial_ma[1:]
+
+        # State covariance matrix
+        if self.k_posdef > 0:
+            self['state_cov', 0, 0] = params_variance
+            if self.state_regression and self.time_varying_regression:
+                self[self._exog_variance_idx] = params_exog_variance
+
+        # Initialize
+        if not self._manual_initialization:
+            self.initialize_state()
+
+        return params
+
+
+class SARIMAXResults(MLEResults):
+    """
+    Class to hold results from fitting an SARIMAX model.
+
+    Parameters
+    ----------
+    model : MLEModel instance
+        The fitted model instance
+
+    Attributes
+    ----------
+    params_included : list of str
+        List of parameters actually included in the model, in sorted order.
```

Comment:
I am going to put the OrderedDict with the slice / index tuple as a TODO rather than in this PR. It's a great idea and would tie in with the `update` method, which needs to know those slices anyway.